### ISSUE TITLE
JdbcSQLException thrown when receiving a transaction with reference-states in Observer node

### ISSUE THREAD

**Issue by berndos** [corda/corda#5178]
When receiving transactions with `subFlow(ReceiveFinalityFlow(otherSideSession = otherSideSession, statesToRecord = StatesToRecord.ALL_VISIBLE))`, `NodeVaultService` throws 
```
[ERROR] 11:36:07,871 [Node thread-3] spi.SqlExceptionHelper.logExceptions - Unique index or primary key violation: "PRIMARY_KEY_9 ON PUBLIC.VAULT_STATES(OUTPUT_INDEX, TRANSACTION_ID) VALUES (0, '4F6535E32DAEC1CA598B4B824B75E1C7040FDF82626B4D36C90F9933D08866EB', 3)"; SQL statement:
insert into vault_states (constraint_data, constraint_type, consumed_timestamp, contract_state_class_name, lock_id, lock_timestamp, notary_name, recorded_timestamp, relevancy_status, state_status, output_index, transaction_id) values (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?) [23505-197] {fiber-id=10000011, flow-id=e43a2a44-348a-4417-a4e6-6ecbf2b0f850, invocation_id=933a20ca-4206-4f9b-a9f6-63cf286458e8, invocation_timestamp=2019-05-29T09:36:07.595Z, origin=O=Alice, L=Zurich, C=CH, session_id=933a20ca-4206-4f9b-a9f6-63cf286458e8, session_timestamp=2019-05-29T09:36:07.595Z, thread-id=431, tx_id=77191E052C057B14F8A1D2BF605ED3A555F959A92284975AED5CEFB431E3B519}
...
[INFO] 11:36:07,893 [Node thread-3] statemachine.StaffedFlowHospital.invoke - Error 1 of 1: {fiber-id=10000011, flow-id=e43a2a44-348a-4417-a4e6-6ecbf2b0f850, invocation_id=933a20ca-4206-4f9b-a9f6-63cf286458e8, invocation_timestamp=2019-05-29T09:36:07.595Z, origin=O=Alice, L=Zurich, C=CH, session_id=933a20ca-4206-4f9b-a9f6-63cf286458e8, session_timestamp=2019-05-29T09:36:07.595Z, thread-id=431, tx_id=77191E052C057B14F8A1D2BF605ED3A555F959A92284975AED5CEFB431E3B519}
 net.corda.core.node.services.VaultQueryException: An error occurred while attempting to query the vault: org.hibernate.exception.ConstraintViolationException: could not execute statement
	at net.corda.node.services.vault.NodeVaultService._queryBy(NodeVaultService.kt:536) ~[corda-node-4.1-RC03.jar:?]
	at net.corda.node.services.vault.NodeVaultService.loadStates(NodeVaultService.kt:807) ~[corda-node-4.1-RC03.jar:?]
	at net.corda.node.services.vault.NodeVaultService.access$loadStates(NodeVaultService.kt:51) ~[corda-node-4.1-RC03.jar:?]
	at net.corda.node.services.vault.NodeVaultService$makeUpdates$4.invoke(NodeVaultService.kt:258) ~[corda-node-4.1-RC03.jar:?]
	at net.corda.node.services.vault.NodeVaultService.makeUpdates(NodeVaultService.kt:325) ~[corda-node-4.1-RC03.jar:?]
	at net.corda.node.services.vault.NodeVaultService.access$makeUpdates(NodeVaultService.kt:51) ~[corda-node-4.1-RC03.jar:?]
	at net.corda.node.services.vault.NodeVaultService$notifyAll$1.invoke(NodeVaultService.kt:212) ~[corda-node-4.1-RC03.jar:?]
	at net.corda.node.services.vault.NodeVaultService.notifyAll(NodeVaultService.kt:223) ~[corda-node-4.1-RC03.jar:?]
...
Caused by: org.h2.jdbc.JdbcSQLException: Unique index or primary key violation: "PRIMARY_KEY_9 ON PUBLIC.VAULT_STATES(OUTPUT_INDEX, TRANSACTION_ID) VALUES (0, '4F6535E32DAEC1CA598B4B824B75E1C7040FDF82626B4D36C90F9933D08866EB', 3)"; SQL statement:
insert into vault_states (constraint_data, constraint_type, consumed_timestamp, contract_state_class_name, lock_id, lock_timestamp, notary_name, recorded_timestamp, relevancy_status, state_status, output_index, transaction_id) values (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?) [23505-197]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:357) ~[h2-1.4.197.jar:1.4.197]
	at org.h2.message.DbException.get(DbException.java:179) ~[h2-1.4.197.jar:1.4.197]

```
This issue can be reproduced with the driver-test in https://github.com/berndos/corda-observer-test (`./gradlew test -i`). I have tested it on `4.0` and `4.1-RC03`.

It appears that resolution of reference-states is not well handled in `NodeVaultService.makeUpdates`, where resolved ref-states are always persisted.  
I can see that there is a check `val notSeenReferences = tx.references - loadStates(tx.references).map { it.ref }` that apparently tries to avoid persisting ref-states that already exist in the vault, but since `loadStates` only returns unconsumed states, and the referenced-state of the tx being resolved might have already been consumed by another tx at the time of resolution, it does not work.

**Comment by mikehearn** [comment]
https://r3-cev.atlassian.net/browse/CORDA-2967

**Comment by nargas-ritu** [comment]
fixed in 4.3
